[Instrumentation.ServiceFabricRemoting] Add RPC metrics instrumentation (#4163)
open-telemetry/opentelemetry-dotnet-contrib: src/OpenTelemetry.Instrumentation.ServiceFabricRemoting/README.md

[![NuGet download count badge](https://img.shields.io/nuget/dt/OpenTelemetry.Instrumentation.ServiceFabricRemoting)](https://www.nuget.org/packages/OpenTelemetry.Instrumentation.ServiceFabricRemoting)
[![codecov.io](https://codecov.io/gh/open-telemetry/opentelemetry-dotnet-contrib/branch/main/graphs/badge.svg?flag=unittests-Instrumentation.ServiceFabricRemoting)](https://app.codecov.io/gh/open-telemetry/opentelemetry-dotnet-contrib?flags[0]=unittests-Instrumentation.ServiceFabricRemoting)

This is an [Instrumentation Library](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/glossary.md#instrumentation-library)

> that instruments [Service Fabric Remoting](https://learn.microsoft.com/en-us/azure/service-fabric/service-fabric-reliable-services-communication-remoting),

Reviewer: ```suggestion
This is an [Instrumentation Library](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/glossary.md#instrumentation-library)
that instruments [Service Fabric Remoting](https://learn.microsoft.com/azure/service-fabric/service-fabric-reliable-services-communication-remoting),
```

        .Build();
```

### Step 4: Metrics

Call the `AddServiceFabricRemotingInstrumentation` extension method on the
`MeterProviderBuilder` to register the OpenTelemetry metrics instrumentation.

```csharp
    using MeterProvider meterProvider = Sdk.CreateMeterProviderBuilder()
        .SetResourceBuilder(ResourceBuilder.CreateDefault().AddService("ServiceFabricRemoting-Example"))
        .AddServiceFabricRemotingInstrumentation()
        .AddOtlpExporter()

>         .Build();

Reviewer: ```suggestion
using MeterProvider meterProvider = Sdk.CreateMeterProviderBuilder()
    .SetResourceBuilder(ResourceBuilder.CreateDefault().AddService("ServiceFabricRemoting-Example"))
    .AddServiceFabricRemotingInstrumentation()
    .AddOtlpExporter()
    .Build();
```

```

Two RPC duration histograms are emitted per the OpenTelemetry

> [RPC metrics semantic conventions](https://opentelemetry.io/docs/specs/semconv/rpc/rpc-metrics/):

Reviewer: Let's point to the GitHub version so we can pin it to the version being implemented:

```suggestion
[RPC metrics semantic conventions](https://github.com/open-telemetry/semantic-conventions/blob/v1.40.0/docs/rpc/rpc-metrics.md):
```

- `rpc.server.call.duration` - histogram of incoming remoting call durations
  (unit: `s`)
- `rpc.client.call.duration` - histogram of outgoing remoting call durations
  (unit: `s`)

Both histograms are tagged with:

- `rpc.system.name` = `"service_fabric_remoting"`
- `rpc.method` - the method name from the remoting request header
- `error.type` - the exception type fully-qualified name, set only when the

>   call throws

Reviewer: What about `rpc.response.status_code` and `server.port`?

Author: SF Remoting doesn't expose a response status code — failures travel as
  serialized exceptions, so there's no natural value for rpc.response.status_code
  (error axis is already covered by error.type).

  For server.port: SF uses dynamic ports that change on failover/restart,
  typically fronted by a reverse proxy (port 19081) and/or Azure Load Balancer.
  The real service port is ephemeral, not visible to most clients, and would
  explode cardinality without adding operational value. Emitting it would also
  conflict with the stable logical identity we set in server.address
  (fabric:/MyApp/MyService), which is what operators actually identify SF
  services by. Skipping server.port is the right call for SF specifically —
  different from gRPC over HTTP, where a stable host:port is meaningful.
  
  I updated the README.md

The client histogram additionally includes `server.address` sourced from
`ResolvedServicePartition.ServiceName` (for example, `"fabric:/MyApp/MyService"`).

The metrics instrumentation has no configuration options: metrics are emitted
automatically for all request/response calls on both Service and Actor
remoting paths. One-way calls (`SendOneWay`) are not recorded, consistent
with the tracing instrumentation.

For consumers that want to register the Meter manually (for example, with a
`MeterListener`), the meter name is exposed as a public constant:

```csharp
string meterName = ServiceFabricRemotingMetrics.MeterName;
// "OpenTelemetry.Instrumentation.ServiceFabricRemoting"

> ```

Reviewer: I think we should either avoid exposing this at all, or make it a `static readonly` instead. Then it's easier to change if needed in the future without recompilation for consumers (e.g. transient dependencies).

Author: removed

## References

> - [Azure Service Fabric documentation](https://learn.microsoft.com/en-us/azure/service-fabric/)

Reviewer: ```suggestion
- [Azure Service Fabric documentation](https://learn.microsoft.com/azure/service-fabric/)
```

The client histogram additionally includes `server.address` sourced from
`ResolvedServicePartition.ServiceName` (for example, `"fabric:/MyApp/MyService"`).

The following RPC semantic convention attributes are intentionally **not**
populated:

- **`rpc.response.status_code`** - Service Fabric Remoting has no native
  status-code concept; failures are serialized exceptions, and the failure
  axis is already captured via the `error.type` tag.
- **`server.port`** - Service Fabric assigns dynamic ports that change on
  failover or restart, and traffic typically flows through a reverse proxy
  or load balancer where the replica port is not visible to callers.
  `server.address` uses the stable logical SF service URI instead, which is

>   how SF services are actually identified in practice.

Reviewer: ```suggestion
  `server.address` uses the stable logical Service Fabric service URI instead, which is
  how Service Fabric services are actually identified in practice.
```
